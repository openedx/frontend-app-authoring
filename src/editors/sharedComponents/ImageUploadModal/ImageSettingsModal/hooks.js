import React from 'react';

import { StrictDict } from '../../../utils';
// This 'module' self-import hack enables mocking during tests.
// See src/editors/decisions/0005-internal-editor-testability-decisions.md. The whole approach to how hooks are tested
// should be re-thought and cleaned up to avoid this pattern.
// eslint-disable-next-line import/no-self-import
import * as module from './hooks';

// Simple wrappers for useState to allow easy mocking for tests.
export const state = {
  // eslint-disable-next-line react-hooks/rules-of-hooks
  altText: (val) => React.useState(val),
  // eslint-disable-next-line react-hooks/rules-of-hooks
  dimensions: (val) => React.useState(val),
  // eslint-disable-next-line react-hooks/rules-of-hooks
  showAltTextDismissibleError: (val) => React.useState(val),
  // eslint-disable-next-line react-hooks/rules-of-hooks
  showAltTextSubmissionError: (val) => React.useState(val),
  // eslint-disable-next-line react-hooks/rules-of-hooks
  isDecorative: (val) => React.useState(val),
  // eslint-disable-next-line react-hooks/rules-of-hooks
  isLocked: (val) => React.useState(val),
  // eslint-disable-next-line react-hooks/rules-of-hooks
  local: (val) => React.useState(val),
  // eslint-disable-next-line react-hooks/rules-of-hooks
  lockAspectRatio: (val) => React.useState(val),
};

export const dimKeys = StrictDict({
  height: 'height',
  width: 'width',
});

/**
 * findGcd(numerator, denominator)
 * Find the greatest common denominator of a ratio or fraction, which may be 1.
 * @param {number} numerator - ratio numerator
 * @param {number} denominator - ratio denominator
 * @return {number} - ratio greatest common denominator
 */
export const findGcd = (a, b) => {
  const gcd = b ? findGcd(b, a % b) : a;

  if (gcd === 1 || [a, b].some(v => !Number.isInteger(v / gcd))) {
    return 1;
  }

  return gcd;
};

const checkEqual = (d1, d2) => (d1.height === d2.height && d1.width === d2.width);

/**
 * getValidDimensions({ dimensions, local, locked })
 * Find valid ending dimensions based on start state, request, and lock state
 * @param {obj} dimensions - current stored dimensions
 * @param {obj} local - local (active) dimensions in the inputs
 * @param {obj} locked - locked dimensions
 * @return {obj} - output dimensions after move ({ height, width })
 */
export const getValidDimensions = ({
  dimensions,
  local,
  isLocked,
  lockAspectRatio,
}) => {
  // if lock is not active, just return new dimensions.
  // If lock is active, but dimensions have not changed, also just return new dimensions.
  if (!isLocked || checkEqual(local, dimensions)) {
    return local;
  }

  const out = {};

  // changed key is value of local height if that has changed, otherwise width.
  const keys = (local.height !== dimensions.height)
    ? { changed: dimKeys.height, other: dimKeys.width }
    : { changed: dimKeys.width, other: dimKeys.height };

  out[keys.changed] = local[keys.changed];
  out[keys.other] = Math.round((local[keys.changed] * lockAspectRatio[keys.other]) / lockAspectRatio[keys.changed]);

  return out;
};

/**
 * reduceDimensions(width, height)
 * reduces both values by dividing by their greates common denominator (which can simply be 1).
 * @return {Array} [width, height]
 */
export const reduceDimensions = (width, height) => {
  const gcd = module.findGcd(width, height);

  return [width / gcd, height / gcd];
};

/**
 * dimensionLockHooks({ dimensions })
 * Returns a set of hooks pertaining to the dimension locks.
 * Locks the dimensions initially, on lock initialization.
 * @param {obj} dimensions - current stored dimensions
 * @return {obj} - dimension lock hooks
 *   {func} initializeLock - enable the lock mechanism
 *   {bool} isLocked - are dimensions locked?
 *   {obj} lockAspectRatio - image dimensions ({ height, width })
 *   {func} lock - lock the dimensions
 *   {func} unlock - unlock the dimensions
 */
export const dimensionLockHooks = () => {
  const [lockAspectRatio, setLockAspectRatio] = module.state.lockAspectRatio(null);
  const [isLocked, setIsLocked] = module.state.isLocked(true);

  const initializeLock = ({ width, height }) => {
    // width and height are treated as a fraction and reduced.
    const [w, h] = reduceDimensions(width, height);

    setLockAspectRatio({ width: w, height: h });
  };

  return {
    initializeLock,
    isLocked,
    lock: () => setIsLocked(true),
    lockAspectRatio,
    unlock: () => setIsLocked(false),
  };
};

/**
 * dimensionHooks()
 * Returns an object of dimension-focused react hooks.
 * @return {obj} - dimension hooks
 *   {func} onImgLoad - initializes image dimension fields
 *     @param {object} selection - selected image object with possible override dimensions.
 *     @return {callback} - image load event callback that loads dimensions.
 *   {object} locked - current locked state
 *   {func} lock - lock current dimensions
 *   {func} unlock - unlock dimensions
 *   {object} value - current dimension values
 *   {func} setHeight - set height
 *     @param {string} - new height string
 *   {func} setWidth - set width
 *     @param {string} - new width string
 *   {func} updateDimensions - set dimensions based on state
 *   {obj} errorProps - props for user feedback error
 *     {bool} isError - true if dimensions are blank
 *     {func} setError - sets isError to true
 *     {func} dismissError - sets isError to false
 *     {bool} isHeightValid - true if height field is ready to save
 *     {func} setHeightValid - sets isHeightValid to true
 *     {func} setHeightNotValid - sets isHeightValid to false
 *     {bool} isWidthValid - true if width field is ready to save
 *     {func} setWidthValid - sets isWidthValid to true
 *     {func} setWidthNotValid - sets isWidthValid to false
 */
export const dimensionHooks = (altTextHook) => {
  const [dimensions, setDimensions] = module.state.dimensions(null);
  const [local, setLocal] = module.state.local(null);

  const setAll = ({ height, width, altText }) => {
    if (altText === '' || altText) {
      if (altText === '') {
        altTextHook.setIsDecorative(true);
      }
      altTextHook.setValue(altText);
    }
    setDimensions({ height, width });
    setLocal({ height, width });
  };

  const setHeight = (height) => {
    if (height.match(/[0-9]+[%]{1}/)) {
      const heightPercent = height.match(/[0-9]+[%]{1}/)[0];
      setLocal({ ...local, height: heightPercent });
    } else if (height.match(/[0-9]/)) {
      setLocal({ ...local, height: parseInt(height, 10) });
    }
  };

  const setWidth = (width) => {
    if (width.match(/[0-9]+[%]{1}/)) {
      const widthPercent = width.match(/[0-9]+[%]{1}/)[0];
      setLocal({ ...local, width: widthPercent });
    } else if (width.match(/[0-9]/)) {
      setLocal({ ...local, width: parseInt(width, 10) });
    }
  };

  const {
    initializeLock,
    isLocked,
    lock,
    lockAspectRatio,
    unlock,
  } = module.dimensionLockHooks({ dimensions });

  return {
    onImgLoad: (selection) => ({ target: img }) => {
      const imageDims = { height: img.naturalHeight, width: img.naturalWidth };
      setAll(selection.height ? selection : imageDims);
      initializeLock(selection.height ? selection : imageDims);
    },
    isLocked,
    lock,
    unlock,
    value: local,
    setHeight,
    setWidth,
    updateDimensions: () => setAll(module.getValidDimensions({
      dimensions,
      local,
      isLocked,
      lockAspectRatio,
    })),
  };
};

/**
 * altTextHooks(savedText)
 * Returns a set of react hooks focused around alt text
 * @return {obj} - alt text hooks
 *   {string} value - alt text value
 *   {func} setValue - set alt test value
 *     @param {string} - new alt text
 *   {bool} isDecorative - is the image decorative?
 *   {func} setIsDecorative - set isDecorative field
 *   {obj} error - error at top of page
 *     {bool} show - is error being displayed?
 *     {func} set - set show to true
 *     {func} dismiss - set show to false
 *   {obj} validation - local alt text error
 *     {bool} show - is validation error being displayed?
 *     {func} set - set validation to true
 *     {func} dismiss - set validation to false
 */
export const altTextHooks = (savedText) => {
  const [value, setValue] = module.state.altText(savedText || '');
  const [isDecorative, setIsDecorative] = module.state.isDecorative(false);
  const [showAltTextDismissibleError, setShowAltTextDismissibleError] = module.state.showAltTextDismissibleError(false);
  const [showAltTextSubmissionError, setShowAltTextSubmissionError] = module.state.showAltTextSubmissionError(false);

  const validateAltText = (newVal, newDecorative) => {
    if (showAltTextSubmissionError) {
      if (newVal || newDecorative) {
        setShowAltTextSubmissionError(false);
      }
    }
  };

  return {
    value,
    setValue: (val) => {
      setValue(val);
      validateAltText(val, null);
    },
    isDecorative,
    setIsDecorative: (decorative) => {
      setIsDecorative(decorative);
      validateAltText(null, decorative);
    },
    error: {
      show: showAltTextDismissibleError,
      set: () => setShowAltTextDismissibleError(true),
      dismiss: () => setShowAltTextDismissibleError(false),
    },
    validation: {
      show: showAltTextSubmissionError,
      set: () => setShowAltTextSubmissionError(true),
      dismiss: () => setShowAltTextSubmissionError(false),
    },
  };
};

/**
 * onInputChange(handleValue)
 * Simple event handler forwarding the event target value to a given callback
 * @param {func} handleValue - event value handler
 * @return {func} - evt callback that will call handleValue with the event target value.
 */
export const onInputChange = (handleValue) => (e) => handleValue(e.target.value);

/**
 * onCheckboxChange(handleValue)
 * Simple event handler forwarding the event target checked prop to a given callback
 * @param {func} handleValue - event value handler
 * @return {func} - evt callback that will call handleValue with the event target checked prop.
 */
export const onCheckboxChange = (handleValue) => (e) => handleValue(e.target.checked);

/**
 * checkFormValidation({ altText, isDecorative, onAltTextFail })
 * Handle saving the image context to the text editor
 * @param {string} altText - image alt text
 * @param {bool} isDecorative - is the image decorative?
 * @param {func} onAltTextFail - called if alt text validation fails
 */
export const checkFormValidation = ({
  altText,
  isDecorative,
  onAltTextFail,
}) => {
  if (!isDecorative && altText === '') {
    onAltTextFail();
    return false;
  }
  return true;
};

/**
 * onSave({ altText, dimensions, isDecorative, saveToEditor })
 * Handle saving the image context to the text editor
 * @param {string} altText - image alt text
 * @param {object} dimensions - image dimensions ({ width, height })
 * @param {bool} isDecorative - is the image decorative?
 * @param {func} saveToEditor - save method for submitting image settings.
 */
export const onSaveClick = ({
  altText,
  dimensions,
  isDecorative,
  saveToEditor,
}) => () => {
  if (module.checkFormValidation({
    altText: altText.value,
    isDecorative,
    onAltTextFail: () => {
      altText.error.set();
      altText.validation.set();
    },
  })) {
    altText.error.dismiss();
    altText.validation.dismiss();
    // Replaces double quotes with &quot; to prevent the alt text from being truncated
    // or breaking the image tag structure.
    const altTextValue = altText.value.replace(/"/g, '&quot;');
    saveToEditor({
      altText: altTextValue,
      dimensions,
      isDecorative,
    });
  }
};
