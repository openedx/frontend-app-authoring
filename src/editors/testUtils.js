/* istanbul ignore file */
import react from 'react';
import { StrictDict } from './utils';

/**
 * Mocked formatMessage provided by react-intl
 */
export const formatMessage = (msg, values) => {
  let message = msg.defaultMessage;
  if (values === undefined) {
    return message;
  }
  Object.keys(values).forEach((key) => {
    // eslint-disable-next-line
    message = message.replace(`{${key}}`, values[key]);
  });
  return message;
};

/**
 * Mock a single component, or a nested component so that its children render nicely
 * in snapshots.
 * @param {string} name - parent component name
 * @param {obj} contents - object of child components with intended component
 *   render name.
 * @return {func} - mock component with nested children.
 *
 * usage:
 *   mockNestedComponent('Card', { Body: 'Card.Body', Form: { Control: { Feedback: 'Form.Control.Feedback' }}... });
 *   mockNestedComponent('IconButton', 'IconButton');
 */
export const mockNestedComponent = (name, contents) => {
  if (typeof contents !== 'object') {
    return contents;
  }
  const fn = () => name;
  Object.defineProperty(fn, 'name', { value: name });
  Object.keys(contents).forEach((nestedName) => {
    const value = contents[nestedName];
    fn[nestedName] = typeof value !== 'object'
      ? value
      : mockNestedComponent(`${name}.${nestedName}`, value);
  });
  return fn;
};

/**
 * Mock a module of components.  nested components will be rendered nicely in snapshots.
 * @param {obj} mapping - component module mock config.
 * @return {obj} - module of flat and nested components that will render nicely in snapshots.
 * usage:
 *   mockNestedComponents({
 *     Card: { Body: 'Card.Body' },
 *     IconButton: 'IconButton',
 *   })
 */
export const mockNestedComponents = (mapping) => Object.entries(mapping).reduce(
  (obj, [name, value]) => ({
    ...obj,
    [name]: mockNestedComponent(name, value),
  }),
  {},
);

/**
 * Mock utility for working with useState in a hooks module.
 * Expects/requires an object containing the state object in order to ensure
 * the mock behavior works appropriately.
 *
 * Expected format:
 *   hooks = { state: { <key>: (val) => React.createRef(val), ... } }
 *
 * Returns a utility for mocking useState and providing access to specific state values
 * and setState methods, as well as allowing per-test configuration of useState value returns.
 *
 * Example usage:
 *   // hooks.js
 *   import * as module from './hooks';
 *   const state = {
 *     isOpen: (val) => React.useState(val),
 *     hasDoors: (val) => React.useState(val),
 *     selected: (val) => React.useState(val),
 *   };
 *   ...
 *   export const exampleHook = () => {
 *     const [isOpen, setIsOpen] = module.state.isOpen(false);
 *     if (!isOpen) { return null; }
 *     return { isOpen, setIsOpen };
 *   }
 *   ...
 *
 *   // hooks.test.js
 *   import * as hooks from './hooks';
 *   const state = new MockUseState(hooks)
 *   ...
 *   describe('state hooks', () => {
 *     state.testGetter(state.keys.isOpen);
 *     state.testGetter(state.keys.hasDoors);
 *     state.testGetter(state.keys.selected);
 *   });
 *   describe('exampleHook', () => {
 *     beforeEach(() => { state.mock(); });
 *     it('returns null if isOpen is default value', () => {
 *       expect(hooks.exampleHook()).toEqual(null);
 *     });
 *     it('returns isOpen and setIsOpen if isOpen is not null', () => {
 *       state.mockVal(state.keys.isOpen, true);
 *       expect(hooks.exampleHook()).toEqual({
 *         isOpen: true,
 *         setIsOpen: state.setState[state.keys.isOpen],
 *       });
 *     });
 *     afterEach(() => { state.restore(); });
 *   });
 *
 * @param {obj} hooks - hooks module containing a 'state' object
 */
export class MockUseState {
  constructor(hooks) {
    this.hooks = hooks;
    this.oldState = null;
    this.setState = {};
    this.stateVals = {};

    this.mock = this.mock.bind(this);
    this.restore = this.restore.bind(this);
    this.mockVal = this.mockVal.bind(this);
    this.testGetter = this.testGetter.bind(this);
  }

  /**
   * @return {object} - StrictDict of state object keys
   */
  get keys() {
    return StrictDict(Object.keys(this.hooks.state).reduce(
      (obj, key) => ({ ...obj, [key]: key }),
      {},
    ));
  }

  /**
   * Replace the hook module's state object with a mocked version, initialized to default values.
   */
  mock() {
    this.oldState = this.hooks.state;
    Object.keys(this.keys).forEach(key => {
      this.hooks.state[key] = jest.fn(val => {
        this.stateVals[key] = val;
        return [val, this.setState[key]];
      });
    });
    this.setState = Object.keys(this.keys).reduce(
      (obj, key) => ({
        ...obj,
        [key]: jest.fn(val => {
          this.hooks.state[key] = val;
        }),
      }),
      {},
    );
  }

  /**
   * Restore the hook module's state object to the actual code.
   */
  restore() {
    this.hooks.state = this.oldState;
  }

  /**
   * Mock the state getter associated with a single key to return a specific value one time.
   * @param {string} key - state key (from this.keys)
   * @param {any} val - new value to be returned by the useState call.
   */
  mockVal(key, val) {
    this.hooks.state[key].mockReturnValueOnce([val, this.setState[key]]);
  }

  testGetter(key) {
    test(`${key} state getter should return useState passthrough`, () => {
      const testValue = 'some value';
      const useState = (val) => ({ useState: val });
      jest.spyOn(react, 'useState').mockImplementationOnce(useState);
      expect(this.hooks.state[key](testValue)).toEqual(useState(testValue));
    });
  }
}
