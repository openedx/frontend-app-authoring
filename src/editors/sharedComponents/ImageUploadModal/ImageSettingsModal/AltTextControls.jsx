import React from 'react';
import PropTypes from 'prop-types';

import { Form } from '@openedx/paragon';
import { FormattedMessage, useIntl } from '@edx/frontend-platform/i18n';

import * as hooks from './hooks';
import messages from './messages';

/**
 * Wrapper for alt-text input and isDecorative checkbox control
 * @param {obj} errorProps - props for error handling
 *   {bool} isValid - are alt-text fields valid for saving?
 * @param {bool} isDecorative - is the image decorative?
 * @param {func} setIsDecorative - handle isDecorative change event
 * @param {func} setValue - update alt-text value
 * @param {string} value - current alt-text value
 */
const AltTextControls = ({
  isDecorative,
  setIsDecorative,
  setValue,
  validation,
  value,
}) => {
  const intl = useIntl();
  return (
    <Form.Group className="mt-4.5">
      <Form.Label as="h4">
        <FormattedMessage {...messages.accessibilityLabel} />
      </Form.Label>
      <Form.Control
        className="mt-4.5"
        disabled={isDecorative}
        floatingLabel={intl.formatMessage(messages.altTextFloatingLabel)}
        isInvalid={validation.show}
        onChange={hooks.onInputChange(setValue)}
        type="input"
        value={value}
      />
      {validation.show
      && (
        <Form.Control.Feedback type="invalid">
          <FormattedMessage {...messages.altTextLocalFeedback} />
        </Form.Control.Feedback>
      )}
      <Form.Checkbox
        checked={isDecorative}
        className="mt-4.5 decorative-control-label"
        onChange={hooks.onCheckboxChange(setIsDecorative)}
      >
        <Form.Label>
          <FormattedMessage {...messages.decorativeAltTextCheckboxLabel} />
        </Form.Label>
      </Form.Checkbox>
    </Form.Group>
  );
};
AltTextControls.propTypes = {
  error: PropTypes.shape({
    show: PropTypes.bool,
  }).isRequired,
  isDecorative: PropTypes.bool.isRequired,
  setValue: PropTypes.func.isRequired,
  setIsDecorative: PropTypes.func.isRequired,
  validation: PropTypes.shape({
    show: PropTypes.bool,
  }).isRequired,
  value: PropTypes.string.isRequired,
};

export const AltTextControlsInternal = AltTextControls; // For testing only
export default AltTextControls;
