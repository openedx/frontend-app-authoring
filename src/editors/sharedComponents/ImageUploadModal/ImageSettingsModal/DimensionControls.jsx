import React from 'react';
import PropTypes from 'prop-types';
import {
  Form,
  Icon,
  IconButton,
} from '@openedx/paragon';
import {
  Locked,
  Unlocked,
} from '@openedx/paragon/icons';
import { FormattedMessage, useIntl } from '@edx/frontend-platform/i18n';

import * as hooks from './hooks';
import messages from './messages';

/**
 * Wrapper for image dimension inputs and the lock checkbox.
 * @param {bool} isLocked - are dimensions locked
 * @param {func} lock - lock dimensions
 * @param {func} setHeight - updates dimensions based on new height
 * @param {func} setWidth - updates dimensions based on new width
 * @param {func} unlock - unlock dimensions
 * @param {func} updateDimensions - update dimensions callback
 * @param {obj} value - local dimension values { height, width }
 */
const DimensionControls = ({
  isLocked,
  lock,
  setHeight,
  setWidth,
  unlock,
  updateDimensions,
  value,
}) => {
  const intl = useIntl();
  if (!value) { return null; }
  return (
    <Form.Group>
      <Form.Label as="h4">
        <FormattedMessage {...messages.imageDimensionsLabel} />
      </Form.Label>
      <div className="mt-4.5">
        <Form.Control
          className="dimension-input"
          value={value.width}
          onChange={hooks.onInputChange(setWidth)}
          onBlur={updateDimensions}
          floatingLabel={intl.formatMessage(messages.widthFloatingLabel)}
        />
        <Form.Control
          className="dimension-input"
          value={value.height}
          onChange={hooks.onInputChange(setHeight)}
          onBlur={updateDimensions}
          floatingLabel={intl.formatMessage(messages.heightFloatingLabel)}
        />
        <IconButton
          className="d-inline-block"
          alt={
          isLocked
            ? intl.formatMessage(messages.unlockDimensionsLabel)
            : intl.formatMessage(messages.lockDimensionsLabel)
        }
          iconAs={Icon}
          src={isLocked ? Locked : Unlocked}
          onClick={isLocked ? unlock : lock}
        />
      </div>
    </Form.Group>
  );
};
DimensionControls.defaultProps = {
  value: {
    height: '100',
    width: '100',
  },
};
DimensionControls.propTypes = ({
  value: PropTypes.shape({
    height: PropTypes.string,
    width: PropTypes.string,
  }),
  setHeight: PropTypes.func.isRequired,
  setWidth: PropTypes.func.isRequired,
  isLocked: PropTypes.bool.isRequired,
  lock: PropTypes.func.isRequired,
  unlock: PropTypes.func.isRequired,
  updateDimensions: PropTypes.func.isRequired,
});

export default DimensionControls;
