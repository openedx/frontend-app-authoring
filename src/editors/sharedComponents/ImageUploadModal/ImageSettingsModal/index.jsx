import React from 'react';
import PropTypes from 'prop-types';

import { Button, Image } from '@openedx/paragon';
import { ArrowBackIos } from '@openedx/paragon/icons';
import { FormattedMessage, useIntl } from '@edx/frontend-platform/i18n';

import './index.scss';
import * as hooks from './hooks';
import messages from './messages';
import BaseModal from '../../BaseModal';
import AltTextControls from './AltTextControls';
import DimensionControls from './DimensionControls';
import ErrorAlert from '../../ErrorAlerts/ErrorAlert';

/**
 * Modal display wrapping the dimension and alt-text controls for image tags
 * inserted into the TextEditor TinyMCE context.
 * Provides a thumbnail and populates dimension and alt-text controls.
 * @param {bool} isOpen - is the modal open?
 * @param {func} close - close the modal
 * @param {obj} selection - current image selection object
 * @param {func} saveToEditor - save the current settings to the editor
 * @param {func} returnToSelection - return to image selection
 */
const ImageSettingsModal = ({
  close,
  isOpen,
  returnToSelection,
  saveToEditor,
  selection,
}) => {
  const intl = useIntl();
  const altText = hooks.altTextHooks(selection.altText);
  const dimensions = hooks.dimensionHooks(altText);
  const onSaveClick = hooks.onSaveClick({
    altText,
    dimensions: dimensions.value,
    isDecorative: altText.isDecorative,
    saveToEditor,
  });
  return (
    <BaseModal
      close={close}
      confirmAction={(
        <Button
          variant="primary"
          onClick={onSaveClick}
        >
          <FormattedMessage {...messages.saveButtonLabel} />
        </Button>
      )}
      isOpen={isOpen}
      title={intl.formatMessage(messages.titleLabel)}
    >
      <ErrorAlert
        dismissError={altText.error.dismiss}
        hideHeading
        isError={altText.error.show}
      >
        <FormattedMessage {...messages.altTextError} />
      </ErrorAlert>
      <Button
        iconBefore={ArrowBackIos}
        onClick={returnToSelection}
        size="inline"
        variant="link"
      >
        <FormattedMessage {...messages.replaceImageButtonLabel} />
      </Button>
      <br />
      <div className="d-flex flex-row m-2 img-settings-form-container">
        <div className="img-settings-thumbnail-container">
          <Image
            className="img-settings-thumbnail"
            onError={dimensions.onImgLoad(selection)}
            onLoad={dimensions.onImgLoad(selection)}
            src={selection.externalUrl}
          />
        </div>
        <hr className="h-100 bg-primary-200 m-0" />
        <div className="img-settings-form-controls">
          <DimensionControls {...dimensions} />
          <AltTextControls {...altText} />
        </div>
      </div>
    </BaseModal>
  );
};

ImageSettingsModal.propTypes = {
  close: PropTypes.func.isRequired,
  isOpen: PropTypes.bool.isRequired,
  returnToSelection: PropTypes.func.isRequired,
  saveToEditor: PropTypes.func.isRequired,
  selection: PropTypes.shape({
    altText: PropTypes.string,
    externalUrl: PropTypes.string,
    url: PropTypes.string,
  }).isRequired,
};
export default ImageSettingsModal;
