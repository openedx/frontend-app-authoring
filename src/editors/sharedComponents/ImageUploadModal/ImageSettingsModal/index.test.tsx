import React from 'react';
import { render, screen, initializeMocks } from '@src/testUtils';
import ImageSettingsModal from '.';

jest.mock('./AltTextControls', () => 'AltTextControls');
jest.mock('./DimensionControls', () => 'DimensionControls');

jest.mock('./hooks', () => ({
  altTextHooks: () => ({
    error: {
      show: true,
      dismiss: jest.fn(),
    },
    isDecorative: false,
    value: 'alternative Taxes',
  }),
  dimensionHooks: () => ({
    onImgLoad: jest.fn(
      (selection) => ({ 'hooks.dimensions.onImgLoad.callback': { selection } }),
    ).mockName('hooks.dimensions.onImgLoad'),
    value: { width: 12, height: 13 },
  }),
  onSaveClick: (args) => ({ 'hooks.onSaveClick': args }),
}));

describe('ImageSettingsModal', () => {
  const props = {
    isOpen: true,
    selection: {
      altText: 'AlTTExt',
      externalUrl: 'ExtERNALurL',
      url: 'UrL',
    },
    close: jest.fn().mockName('props.close'),
    returnToSelection: jest.fn().mockName('props.returnToSelector'),
    saveToEditor: jest.fn().mockName('props.saveToEditor'),
  };
  beforeEach(() => {
    initializeMocks();
  });
  test('renders component', () => {
    render(<ImageSettingsModal {...props} />);
    expect(screen.getByText('Image Settings')).toBeInTheDocument();
  });
});
