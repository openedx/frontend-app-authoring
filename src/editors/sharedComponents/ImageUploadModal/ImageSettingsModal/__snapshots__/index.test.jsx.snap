// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageSettingsModal render snapshot 1`] = `
<BaseModal
  bodyStyle={null}
  close={[MockFunction props.close]}
  confirmAction={
    <Button
      onClick={
        {
          "hooks.onSaveClick": {
            "altText": {
              "error": {
                "dismiss": [MockFunction],
                "show": true,
              },
              "isDecorative": false,
              "value": "alternative Taxes",
            },
            "dimensions": {
              "height": 13,
              "width": 12,
            },
            "isDecorative": false,
            "saveToEditor": [MockFunction props.saveToEditor],
          },
        }
      }
      variant="primary"
    >
      <FormattedMessage
        defaultMessage="Save"
        description="Label for save button."
        id="authoring.texteditor.imagesettingsmodal.saveButtonLabel"
      />
    </Button>
  }
  footerAction={null}
  headerComponent={null}
  hideCancelButton={false}
  isFullscreenScroll={true}
  isOpen={false}
  size="lg"
  title="Image Settings"
>
  <ErrorAlert
    dismissError={[MockFunction]}
    hideHeading={true}
    isError={true}
  >
    <FormattedMessage
      defaultMessage="Enter alt text or specify that the image is decorative only."
      description="Message presented to user when user attempts to save unaccepted altText configuration."
      id="authoring.texteditor.imagesettingsmodal.error.altTextError"
    />
  </ErrorAlert>
  <Button
    onClick={[MockFunction props.returnToSelector]}
    size="inline"
    variant="link"
  >
    <FormattedMessage
      defaultMessage="Replace image"
      description="Label for replace image button."
      id="authoring.texteditor.imagesettingsmodal.replaceImageButtonLabel"
    />
  </Button>
  <br />
  <div
    className="d-flex flex-row m-2 img-settings-form-container"
  >
    <div
      className="img-settings-thumbnail-container"
    >
      <Image
        className="img-settings-thumbnail"
        onError={
          {
            "hooks.dimensions.onImgLoad.callback": {
              "selection": {
                "altText": "AlTTExt",
                "externalUrl": "ExtERNALurL",
                "url": "UrL",
              },
            },
          }
        }
        onLoad={
          {
            "hooks.dimensions.onImgLoad.callback": {
              "selection": {
                "altText": "AlTTExt",
                "externalUrl": "ExtERNALurL",
                "url": "UrL",
              },
            },
          }
        }
        src="ExtERNALurL"
      />
    </div>
    <hr
      className="h-100 bg-primary-200 m-0"
    />
    <div
      className="img-settings-form-controls"
    >
      <DimensionControls
        onImgLoad={
          [MockFunction hooks.dimensions.onImgLoad] {
            "calls": [
              [
                {
                  "altText": "AlTTExt",
                  "externalUrl": "ExtERNALurL",
                  "url": "UrL",
                },
              ],
              [
                {
                  "altText": "AlTTExt",
                  "externalUrl": "ExtERNALurL",
                  "url": "UrL",
                },
              ],
            ],
            "results": [
              {
                "type": "return",
                "value": {
                  "hooks.dimensions.onImgLoad.callback": {
                    "selection": {
                      "altText": "AlTTExt",
                      "externalUrl": "ExtERNALurL",
                      "url": "UrL",
                    },
                  },
                },
              },
              {
                "type": "return",
                "value": {
                  "hooks.dimensions.onImgLoad.callback": {
                    "selection": {
                      "altText": "AlTTExt",
                      "externalUrl": "ExtERNALurL",
                      "url": "UrL",
                    },
                  },
                },
              },
            ],
          }
        }
        value={
          {
            "height": 13,
            "width": 12,
          }
        }
      />
      <AltTextControls
        error={
          {
            "dismiss": [MockFunction],
            "show": true,
          }
        }
        isDecorative={false}
        value="alternative Taxes"
      />
    </div>
  </div>
</BaseModal>
`;
